Pipeline sharing support for alternative BitBucket Server #1297

piotr-faba-ardigen opened the pull request:
New feature as discussed in: https://github.com/nextflow-io/nextflow/issues/1297

Signed-off-by: Piotr Faba <piotr.faba@ardigen.com>

pditommaso commented on modules/nextflow/src/main/groovy/nextflow/scm/BitbucketServerRepositoryProvider.groovy:
```diff
@@ -0,0 +1,87 @@
+ * Implements a repository provider for the BitBucket service
+ *
+ * @author Paolo Di Tommaso <paolo.ditommaso@gmail.com>
+ */
```
The should be your name as the author!

piotr-faba-ardigen commented:
> Thanks for submitting this PR! it looks good, if possible would be nice to have some more tests and the relative documentation. Also for the PR to get approved you will need to agree with the DCO https://github.com/apps/dco.
> 
> Thanks!

Thank you for a quick check. I've updated the PR and added explanations here. Let me know if any further action is required.

1. The author for new files was modified to piotr.faba@ardigen.com
2. Documentation is added in the `Private server configuration` section of the `sharing.rst`. Does this look good?
3. What additional tests would be expected? I implemented the same ones, as for `gitea`. Other public functions would require functioning server. I unfortunately do not have working knowledge of any publicly available server, that could be used for this purpose. Atlassian allows only 30 days evaluation period for this product [product](https://bitbucket.org/product/download). I have tested the functionality with the internally available server that for security reasons cannot be made public.
4. I think I fulfill all DCO requirements (the icon in pull request shows it's OK). Is there anything missing?

pditommaso commented on modules/nextflow/src/main/groovy/nextflow/scm/RepositoryProvider.groovy:
```diff
@@ -153,6 +153,12 @@ abstract class RepositoryProvider {
+                int indexNearUrlStart = conString.indexOf("//") - 10
+                int startIndex = conString.indexOf("http",indexNearUrlStart)
+                throw new AbortOperationException("Uri not found. This uri was attempted for api query: ${conString.substring(startIndex)}")
         }
```
Hi, sorry for the late reply. Was wondering what's the rationale of this snippet. Could you please provide more details?

piotr-faba-ardigen replied:
Good catch! If you refer to this fizzy implementation, I replaced it with `connection.getURL()` in the recent commit.

piotr-faba-ardigen replied:
If you refer to adding support for 404 error, I don't consider it a must have, but it can be a nice option. Without this the message displayed to user is:
```
Cannot find `test_proj` -- Make sure exists a BitBucketServer repository at this address `https://bitbucket.server.com/scm/test_proj/test_repo`
```
This is misleading when dealing with unsupported repository provider, as we new this repository existed, so the information was of no use in such case.

The new message indicates clearer what the source of error is:
```
Uri not found. This uri was attempted for api query: https://bitbucket.biscrum.com/rest/api/1.0/projects/test_proj/repos/test_repo/raw/main.nf
```

pditommaso replied:
I'm ok with that, but it should go in separate PR since it modified a common behaviour.

piotr-faba-ardigen replied:
OK. I'll remove this from this branch and create a new PR.

piotr-faba-ardigen replied:
A separate pull request is created: https://github.com/nextflow-io/nextflow/pull/1341

pditommaso commented:
I'm a bit confused by the BitBucket offering. Could you please link the Bicket product corresponding to this api ?

piotr-faba-ardigen commented:
> I'm a bit confused by the BitBucket offering. Could you please link the Bicket product corresponding to this api implementation?

Hello. I was out of office - hence the delayed reply. The following is the link to BitBucket Server api documentation: https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html

piotr-faba-ardigen commented:
> r users it should be documented how to configure the server credentials. Please add a section as f

I added the Warning in the `sharing.rst` in `Private server configuration` section. Would you prefer than I move it to `BitBucket credentials` section or create a separate one `BitBucket Server credentials`?

pditommaso commented:
I've merged this manually changing a bit the docs. See 6af6a902. Thanks a lot for this PR.

piotr-faba-ardigen commented:
> I've merged this manually changing a bit the docs. See [6af6a90](https://github.com/nextflow-io/nextflow/commit/6af6a90268a85f0f44d235d6e6adf459c88722e9). Thanks a lot for this PR.

Thank you as well
